server.py

```diff
@@ -36,17 +36,25 @@ def index():
 ######################################################################
 @app.route('/products', methods=['GET'])
 def list_products():
-    category        = request.args.get('category')
-    discontinued    = str2bool(request.args.get('discontinued'))
-    results     = []
-    for key, value in products.iteritems():
-        if matchesClause(category,      value['category']) and \
-           matchesClause(discontinued,  value['discontinued']):
-             results.append(products[key])
-    return reply(results, HTTP_200_OK)
+	category=request.args.get('category')
+	discontinued = str2bool(request.args.get('discontinued'))
+	search_term=(request.args.get('search')).lower()
+	#print ("Category", category)
+	#print ("discontinued", discontinued)
+	#print ("search_term", search_term)
+	results = []
+	for key, value in products.iteritems():
+		if category is None and discontinued is None and search_term is None:
```

Comment:
The `matchesClause()` helper method was intended to eliminate long lists of `and` and `or` conditions in `if` statements such as this one on line 47.

The approach of having a helper method for each URL parameter allows us to modularize and encapsulate the logic needed for each of those parameters.

I would therefore suggest and adding another helper method to handle the search term, e.g. `matches_term()` that checks if the `search` parameter is `None`, and if not, check if the product name, category, and description match the search term(s).

So the code from lines 47-53 could be condensed and simplified into something like this:
```
if matches_Clause(category, value['category']) \
and matches_Clause(discontinued, value['discontinued']) \
and matches_search_term(search_term, value):
        results.append(products[key])
```